### ISSUE TITLE
Add Basic CI/CD

### ISSUE THREAD

**Issue by yonipeleg33** [iximiuz/pq#3]
Hello there!

1. Currently, running `cargo clippy` yields:
```
...
error: aborting due to 4 previous errors; 56 warnings emitted
```
2. No formatting is enforced on new PRs.
3. Currently, there's no way of using `pq` without building it locally afaik, which can be a huge drawback.

### Suggestion:
1. Fix all clippy warnings / errors
2. Add a GitHub Action that runs `cargo clippy`, `cargo fmt`, `make test-all` and `make test-e2e` on every PR and push to master
3. Add a GitHub Action that publishes to crates.io so that one can install it using `cargo install pq` (and consider packaging it to `apt` and such, I offered crates.io as it's the easiest option I think)

**Comment by iximiuz** [comment]
Seems like the crate name [`pq`](https://crates.io/crates/pq) is already taken. I added a workaround `cargo install --git https://github.com/iximiuz/pq` to README, but I'd appreciate any suggestions of a better solution.

For the GitHub actions part, I added `cargo fmt` and `cargo test`. I'm looking how to add the `clippy` check.

**Comment by yonip23** [comment]
Mmm idk what to do with the existing name, maybe `pq-rs` is free?

About the clippy, it should be quite trivial to do it in gh actions, lmk if you need any help

**Comment by iximiuz** [comment]
Ok, we have all the basic checks in the action now - `fmt`, `test`, and `clippy`.

**Comment by yonip23** [comment]
👑
All that's left is the crates.io issue then?

**Comment by iximiuz** [comment]
Yeah, although I'm still not sure about the crate name. `pq-rs` sounds good, but not ideal 😃

**Comment by yonip23** [comment]
Btw Im afk now, so if you want to try and run `cargo fmt/clippy --all-targets`, I'm not sure that all of the code is covered by default (examples, tests etc.)

**Comment by yonip23** [comment]
> Yeah, although I'm still not sure about the crate name. `pq-rs` sounds good, but not ideal 😃 

I get you, but it's quite classic to add `rs` to crate names. There are tons of crates with this postfix. Your call of course 🙂

**Comment by iximiuz** [comment]
Indeed, `cargo clippy` doesn't check tests by default. Updated the action yaml.

For the name, I need to sleep on it first 😄

**Comment by yonip23** [comment]
Sure

While you're at it, would be great if you could open another issue or two with a description of what needs to be done (and maybe some hints where to start)😊

**Comment by iximiuz** [comment]
Just added one #6. I have plenty of ideas but most of them are too vague for a good first issue. I'll see how I could formalize and decompose them.

**Comment by yonip23** [comment]
Nice, will look into it soon :)
